diff --git a/tools/xmlschema.py b/tools/xmlschema.py
index b3518f9e..4b342255 100755
--- a/tools/xmlschema.py
+++ b/tools/xmlschema.py
@@ -101,14 +101,14 @@ def print_documentation(element: ElementTree.Element) -> List[str]:
     return lines
 
 
-def print_include(element: ElementTree.Element) -> List[str]:
+def print_include(element: ElementTree.Element, install_dir: str | None = None) -> List[str]:
     """
     Print include tag information
     """
     lines = []
     filename = get_attribute(element, "filename")
     if filename is not None:
-        loc = "http://sdformat.org/schemas/"
+        loc = "file://" + install_dir + "/" if install_dir is not None else "http://sdformat.org/schemas/"
         loc += filename.replace(".sdf", ".xsd")
         lines.append(f"<xsd:include schemaLocation='{loc}'/>")
     return lines
@@ -147,7 +147,7 @@ def print_plugin_element(element: ElementTree.Element) -> List[str]:
     return lines
 
 
-def print_element(element: ElementTree.Element) -> List[str]:
+def print_element(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
     """
     Print a child element of the sdf definition
     """
@@ -173,7 +173,10 @@ def print_element(element: ElementTree.Element) -> List[str]:
         lines.append("    <xsd:choice maxOccurs='unbounded'>")
 
         for child_element in element.findall("element"):
-            lines.extend(indent_lines(print_element(child_element), 6))
+            lines.extend(indent_lines(print_element(child_element, sdf_root_dir), 6))
+
+        for include_element in element.findall("include"):
+            lines.extend(indent_lines(print_include_ref(include_element, sdf_root_dir), 6))
 
         lines.append("    </xsd:choice>")
 
@@ -222,7 +225,7 @@ def print_attribute(element: ElementTree.Element) -> List[str]:
     return lines
 
 
-def print_xsd(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
+def print_xsd(element: ElementTree.Element, sdf_root_dir: str, install_dir: str | None = None) -> List[str]:
     """
     Print xsd for top level SDF element
     """
@@ -233,16 +236,20 @@ def print_xsd(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
 
     elements = element.findall("element")
     attributes = element.findall("attribute")
-    includes = element.findall("include")
+    includes = element.findall(".//include")
 
     lines.extend(print_documentation(element))
     lines.append(
-        "<xsd:include schemaLocation='http://sdformat.org/schemas/types.xsd'/>"
+        "<xsd:include schemaLocation='{loc}/types.xsd'/>".format(loc="file://" + install_dir if install_dir is not None else "http://sdformat.org/schemas/")
     )
 
     # Reference any includes in the SDF file
+    included = []
     for include in includes:
-        lines.extend(print_include(include))
+        filename = get_attribute(include, "filename")
+        if filename not in included:
+            lines.extend(print_include(include, install_dir))
+            included.append(filename)
 
     if len(elements) or len(attributes) or len(includes):
         lines.append(f"<xsd:element name='{elem_name}'>")
@@ -251,12 +258,20 @@ def print_xsd(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
         if elem_name != "plugin" and (len(elements) or len(includes)):
             lines.append("    <xsd:choice maxOccurs='unbounded'>")
 
+        if elem_type is not None:
+            if (len(elements) or len(includes)):
+                lines.append("  <xsd:complexContent>")
+                lines.append(f"  <xsd:extension base='{elem_type}'>")
+            else:
+                lines.append("  <xsd:simpleContent>")
+                lines.append(f"  <xsd:extension base='{elem_type}'>")
+
         for child_element in elements:
             if "copy_data" in child_element.attrib:
                 element_lines = print_plugin_element(child_element)
                 lines.extend(indent_lines(element_lines, 4))
             else:
-                element_lines = print_element(child_element)
+                element_lines = print_element(child_element, sdf_root_dir)
                 lines.extend(indent_lines(element_lines, 6))
 
         for include_element in includes:
@@ -269,6 +284,13 @@ def print_xsd(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
         for attribute_element in attributes:
             lines.extend(indent_lines(print_attribute(attribute_element), 4))
 
+        if elem_type is not None:
+            if (len(elements) or len(includes)):
+                lines.append("  </xsd:extension>")
+                lines.append("  </xsd:complexContent>")
+            else:
+                lines.append(f"  </xsd:extension>")
+                lines.append("  </xsd:simpleContent>")
         lines.append("  </xsd:complexType>")
         lines.append("</xsd:element>")
     else:
@@ -281,7 +303,7 @@ def print_xsd(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
     return lines
 
 
-def process(input_file_sdf: str, sdf_dir: str) -> List[str]:
+def process(input_file_sdf: str, sdf_dir: str, install_dir: str) -> List[str]:
     """
     Produce an XSD file from an input SDF file
     """
@@ -290,7 +312,7 @@ def process(input_file_sdf: str, sdf_dir: str) -> List[str]:
     root = tree.getroot()
     lines.append("<?xml version='1.0' encoding='UTF-8'?>")
     lines.append("<xsd:schema xmlns:xsd='http://www.w3.org/2001/XMLSchema'>")
-    lines.extend(indent_lines(print_xsd(root, sdf_dir), 2))
+    lines.extend(indent_lines(print_xsd(root, sdf_dir, install_dir), 2))
     lines.append("</xsd:schema>")
     return lines
 
@@ -302,10 +324,11 @@ if __name__ == "__main__":
     parser.add_argument("--input-file")
     parser.add_argument("--sdf-dir")
     parser.add_argument("--output-dir")
+    parser.add_argument("--install-dir")
     args = parser.parse_args()
 
     input_file = os.path.abspath(args.input_file)
-    output_lines = process(input_file, args.sdf_dir)
+    output_lines = process(input_file, args.sdf_dir, args.install_dir)
     fname = os.path.splitext(os.path.basename(args.input_file))[0]
     os.makedirs(args.output_dir, exist_ok=True)
 
